### PR TITLE
Update to branch-head/58 and add ARM support.

### PR DESCRIPTION
**Pull request by tomtom5152**
Quite a bit going on in this, my original idea was just to add support for ARM based devices (specifically the [BeagleBone Black](https://beagleboard.org/black) for my use case). This got a bit out of hand when the gyp build no longer support cross compiling (couldn't figure out why). Easiest way to overcome this was to bump WebRTC to the latest version and switch to using `gn gen` to create the ninja files.

As such this PR contains the following changes.

* Add ARM support to build script.
* Update build script commit hash to latest version.
* Add arm pkg-config file.
* Update WebRTC to branch-head/58.
  * Update WebRTC headers.
* Update tests to reflect breaking changes to the WebRTC API (specifically peer connections).
* Add new required methods in the fake data channel.

### CODE REVIEW

**Comment by tomtom5152** [comment]
I should also add that although I'm not sure how to get travis to test this, it has been tested using [resin.io](https://resin.io) on a real device and functions perfectly.

**Comment by arlolra** [comment]
Thanks!  This looks promising.

@keroserene, if you don't mind, I can try and get this merged.

File: build.sh [hunk 1]
```diff
@@ -67,19 +74,15 @@ pushd $WEBRTC_SRC || exit 1
...
 pushd $WEBRTC_SRC
 export GYP_DEFINES="include_tests=0 include_examples=0"
-python webrtc/build/gyp_webrtc webrtc/api/api.gyp || exit 1
-ninja -C out/$CONFIG || exit 1
+if [ "$ARCH" = "arm" ]; then
+    gn gen out/$CONFIG --args='target_os="linux" target_cpu="arm" is_debug=false' || exit 1
+else
+    gn gen out/$CONFIG --args='is_debug=false' || exit 1
+fi
+ninja -C out/$CONFIG webrtc webrtc/test webrtc/pc:pc_test_utils || exit 1
```
Comment (arlolra) [inline build.sh]: I think `ninja -C out/$CONFIG webrtc field_trial metrics_default fake_audio_device` is sufficient.

Comment (arlolra) [inline build.sh]: s/fake_audio_device/pc_test_utils/

**Comment by keroserene** [comment]
@tomtom5152 This looks great, thanks for adding ARM support and updating all the things!
The travis build looks happy at the moment. @arlolra Sounds good, thanks -- so long as it continues to work with snowflake I'd be stoked to see this merged :+1:

**Comment by arlolra** [comment]
Alright, I merged the branch-heads/58 part of this in 9c40f4d6 and efd9a1a8, rebuilt the magic in ab1b6486, and opened #63 for fixing up the configuration methods.

@keroserene I tested snowflake with the above and it bootstrapped tor to 100%.  We're gonna need a corresponding tor-browser-bundle patch, which I'll work on next.

Then I'll get back to merging the arm parts here.

Thanks again @tomtom5152!

**Comment by tomtom5152** [comment]
Thats some git-foo magic going on right there, are you able to test with ARM or would you like me to do it. I also notice that the arm magic didn't get rebuilt here so would you like a new one of those?

**Comment by arlolra** [comment]
@tomtom5152 I haven't merged anything related to arm yet.  I'll be getting that soon.  Thanks for the offer but we probably wouldn't take any binary patches (hence, why I recompiled the magic).

**Comment by tomtom5152** [comment]
Ah fair enough, I can appreciate not taking binary patches, although it wasn't something I had considered. If it could make an appearance soonish though I would be very grateful. 

You've probably figured it out, but the build command is `GOARCH=arm ./build.sh`. I used an AWS EC2 m4.large instance with arm-linux-gnueabi-hf to do it because I didn't have access to a linux machine locally, but your milage might vary.

**Comment by arlolra** [comment]
> We're gonna need a corresponding tor-browser-bundle patch, which I'll work on next.

That's done at, https://trac.torproject.org/projects/tor/ticket/21748#comment:3

Going to work on merge the arm parts now.

File: build.sh [hunk 2]
```diff
@@ -57,6 +57,13 @@ else
 	popd
 fi
 
+if [ "$ARCH" = "arm" ]; then
+    echo "Manually fetching arm sysroot"
```
Comment (arlolra) [inline build.sh]: Why was doing this manually necessary?  `gclient sync` seems to do,

```
________ running '/usr/bin/python src/build/linux/sysroot_scripts/install-sysroot.py --running-as-hook' in '/Sandbox/go-webrtc/third_party/webrtc'
Wheezy arm sysroot image already up to date: /Sandbox/go-webrtc/third_party/webrtc/src/build/linux/debian_wheezy_arm-sysroot
Precise amd64 sysroot image already up to date: /Sandbox/go-webrtc/third_party/webrtc/src/build/linux/ubuntu_precise_amd64-sysroot
```

Comment (arlolra) [inline build.sh]: Gah, sorry, nevermind.  I see you're cross-compiling.

**Comment by arlolra** [comment]
@tomtom5152 Can you test the `arm` branch and confirm if it works for you? (I merged your patch there and build the magic.)  I only have access to an old armv6 pi and setting `--args='arm_version=6 ...'` seems to have bit rotted, since they don't test that in their matrix.

**Comment by tomtom5152** [comment]
@arlolra That works perfectly for me on my BeagleBone, `go test` passes, as does the demo chat system. If you absolutely want to I might be able to set one up on a non standard SSH port for a while to test?

**Comment by arlolra** [comment]
> If you absolutely want to I might be able to set one up on a non standard SSH port for a while to test?

No, that's ok.  I merged the branch as is.  Think we got everything here now.  Thanks for all your help.

**Comment by circuitry2** [comment]
@tomtom5152 Is this support for ARM64 or ARM that you added? I am trying get this running on an ARMV7l machine, but when I run go build I get
`../../../github.com/keroserene/go-webrtc/ctestenums.cc:2:48: fatal error: webrtc/api/peerconnectioninterface.h: No such file or directory
 #include "webrtc/api/peerconnectioninterface.h"`

I believe I successfully cross-compiled for ARM V7 using this setting while building:
`export GYP_DEFINES="OS=linux target_arch=arm arm_version=7 arm_use_neon=1"`

I have a newly generated libwebrtc.a file, and libwebrtc_common.a.

Do you know how I can use these .a files to get go-webrtc working on ARM v7?
Happy to share the files and add support for 32 bit ARM once I get it working.
Thanks!

**Comment by circuitry2** [comment]
I tried swapping out the arm binary with one I built myself, as well as another prebuild arm v7 binary I found online and I'm still seeing the same error.

I made sure that the webrtc-linux-arm.pc script is getting called successfully, and I changed the name of the binary here:
`	-lwebrtc-linux-arm-magic \`

No matter what I do whenever I build I get the same error:
`../../../github.com/keroserene/go-webrtc/ctestenums.cc:2:48: fatal error: webrtc/api/peerconnectioninterface.h: No such file or directory #include "webrtc/api/peerconnectioninterface.h"`

Are my binaries bad, or am I just doing something wrong?
I'm trying to build the project using a vagrant VM that I have that is specifically set up to cross compile go projects for ARM. I haven't had any problems with the VM to date. The VM just says various CGO and GO cross compiling environment variables.

**Comment by circuitry2** [comment]
@tomtom5152 do you have any ideas for how I can use my WebRTC binary with this project?

**Comment by tomtom5152** [comment]
Apologies @circuitry2, I've been mostly offline since just before Christmas and just going through my emails now. My CGO is also rather rusty at this point as I haven't had to argue with it since this PR, I've just been using Go.

By the looks of it you have issues with the include directory config, I'm not sure if this in itself indicates a good binary or not, as I believe it may come before the binary is used.
If this is the case, then it may have something to do with the .pc files in the package root, see my [webrtc-linux-arm.pc](https://github.com/keroserene/go-webrtc/blob/master/webrtc-linux-arm.pc) for more details. You will also need to inform CGO to use this file as per [datachannel.go#12](https://github.com/keroserene/go-webrtc/blob/b40f274c23bbd6c57ab148505297b3f6cdad703a/datachannel.go#L12), although I can't help you with the exact syntax as I have never used ARM64.

That should solve the include issues, but you might now have an issue with the binary. I would strongly suggest you checkout [build.sh](https://github.com/keroserene/go-webrtc/blob/master/build.sh) for how to create a standard binary without all the bloat that the WebRTC instructions give. Particularly [L77](https://github.com/keroserene/go-webrtc/blob/9dfd137319e5846f61cddb2d441d2b8e92ce063c/build.sh#L77) would be of some interest, as the latest [install-sysroot.py](https://chromium.googlesource.com/chromium/src/build/+/master/linux/sysroot_scripts/install-sysroot.py) supports ARM64, and [L117](https://github.com/keroserene/go-webrtc/blob/9dfd137319e5846f61cddb2d441d2b8e92ce063c/build.sh#L117) would be a good starting point for the compile command.
I would also advise that you ensure you are compiling the correct version of WebRTC, this library is a tad behind these days [L12](https://github.com/keroserene/go-webrtc/blob/9dfd137319e5846f61cddb2d441d2b8e92ce063c/build.sh#L12) has the version that I used successfully, although I'm sure a bump the lates WebRTC wouldn't go a miss with this repo, but baby steps.

Hope that's of some use, although I have no use for it at the moment, this is of some interest to me still as I may use ARM64 in the future.

**Comment by uumaro** [comment]
> Is this support for ARM64 or ARM that you added? I am trying get this running on an ARMV7l machine, but when I run go build I get
`../../../github.com/keroserene/go-webrtc/ctestenums.cc:2:48: fatal error: webrtc/api/peerconnectioninterface.h: No such file or directory #include "webrtc/api/peerconnectioninterface.h"`

@circuitry2 This error message is the same as at #76. It probably means that your build is missing the necessary `-Iinclude/` option to refer to the correct header files.

Or, if you built your own webrtc that's newer than branch-heads/58, the problem may be that upstream, the header files [lost a "webrtc/" prefix](https://webrtc.googlesource.com/src/+/bb547203bfebcc478b263c4e9ca173c6fd5a0c5d). So "webrtc/api/peerconnectioninterface.h" needs to become "api/peerconnectioninterface.h" in ctestenums.cc and other places. Your best bet is to build your webrtc using the current commit hash in build.sh.

**Pull request closed without merging**